FATIP-300-FAT-RPC-API-Standard

drkatz opened the pull request:
Initial standard defines the following methods:

**Token get**
- Get token issuance
- Get transaction
- Get transaction history (paged)
- Get balance
- Get stats

**Token write**
- Write Transaction
- Delete Token (un track token from daemon & delete data)

**Daemon get**
- Get properties
- Get tracked tokens

**RPC Error codes**
- See branch

AdamSLevy commented on fatips/300.md:
```diff
@@ -0,0 +1,488 @@
+| `token-id`   | string | Token ID                                                     |                                                              | Y        |
+| `issuer-id`  | string | Issuer Root Chain ID                                         | Must be a valid issuer root chain ID. When combined as per FATIP-100 implementation, must resolve to a valid token chain | Y        |
+| `fa-address` | string | The Factoid address to get transactions for                  | Must be a valid Factoid address. If omitted the transactions to return will be from the set of all token transactions. | N        |
+| `tx-id`      | string | The tx ID to take as the starting point for the next page (exclusive of tx `tx-id`) | Must be a valid FAT tx in the database. If omitted the results are the most recent 25 transactions. If `fa-address` is specified, then the tx must be a historical transaction of `fa-address` | N        |
```
I agree with Paul. Tx-ids are random hashes. You can't easily index into the middle of them without iterating through to learn new tx-ids to request as the starting point. We should allow looking by tx-id and also simply view this as pagination. i.e. add `length` and `page` or something like that. If I want the 25 tx's after the first 50 I ask for page 3 with length 25.

AdamSLevy commented:
Some additional comments:
- Let's allow the user to look up by the chain id as well as the token-id + issuer-id.
- It might be nice to be able to look up, "what are my non-zero balances for this address for any FAT tokens fatd knows about" so a query across all tokens.
- Im not sure we should have an API call for deleting a token's data. Since a token is stored in a database, the user can just remove the database. maybe we add a blacklist/whitelist feature to stop fatd from tracking certain tokens. I don't see a strong need to give this much control to an API.

- Awesome work on the error codes. I may need to rework my jsonrpc lib to allow those through. But it probably should if it doesn't... Need to double check that...

drkatz commented:
Thanks for the comments @AdamSLevy, will address these in a upcoming commit.

drkatz commented:
@AdamSLevy The above commit [3bad877](https://github.com/Factom-Asset-Tokens/FAT/pull/20/commits/3bad877ab71dde2bfc6c374b28ebb1ec9cd2e2a9):

- Addresses your points 1 and 3 (remove `delete-token` and add `chain-id`)
- Streamlines the requirement for `token-id` & `issuer-id` OR `chain-id`
- Adds ExtIds to responses by the RPC endpoint for transactions and issuances.

For now I'll hold out on the method to get nonzero balances for an address among tracked tokens, as we should strategize a bit. How expensive do you envision this being? Off the top of my head, given a list of addresses it's pretty efficient to check the balance for each token (It's a key value relationship). What are your thoughts?

drkatz commented:
I think the spec should note just to be explicit, that specifying `chain-id` is required to be either the tx or issuance chain. Either is fine since we're just trying to find the token ID & Issuer from ExtIds and either will provide that information.

nklomp commented:
Warning beforehand. I started at the RPC spec, so some questions/remarks might be of ;)

> get-issuance:
- in the request you have a issuer-id whilst in the response it is "issuer"
- why is token-id not returned (I know you supply it already)? 

> get-transaction:
- Is there are reason I have to provide the token-id and issuer-id here? The tx-id is globally unique right? It is mandatory for almost all other methods, so that might be a symmetry reason
- tx-id vs txId

> get-transactions:
- fa-address: Input or output or both?
- why can't I control pagination myself?

> delete-token:
Although only local, don't you want to sign this somehow?

> get-daemon-tokens vs delete-token
Daemon in one but not in other


A HTTP 400 code for invalid address is not what I would expect from a REST perspective. I would expect another 404 (depends a bit on what you define as invalid address though) Because at least you also want an address not found

I would suggest HTTP 409 for Token Syncing. 408 might mean people actually think you want to close a connection and expect corresponding header fields ;)

drkatz commented:
> in the request you have a issuer-id whilst in the response it is "issuer"

Actually "issuer' isn't even returned in the response according to the PR that was merged to master  earlier today(#15). The library implementing the RPC should splice in the issuer root chain ID if desired.

> why is token-id not returned (I know you supply it already)?

Because it already has to be known to make the call in the first place :wink: 

> Is there are reason I have to provide the token-id and issuer-id here? 

Search space efficiency & database structure. The database on the FAT Daemon is structured such that each FAT token has its own database, and it's own database file as a consequence. This database separation allows easier debugging and portability, and gets around column modularity(this is sqlite not nosql). As a consequence, the number of transactions to be queried over is lower since the tx tables are separated, but you need to supply token-id and issuer-id.

> tx-id vs txId

:potato: :tomato: I'm open to renaming this thing, but then we'd need to rename lots of other things :wink: 

> fa-address: Input or output or both?

Both. Thanks for pointing this out, the spec should be more specific. Perhaps in the future we can allow choosing.

> why can't I control pagination myself?

Could you please explain how you are limited in your control of pagination with the current API?

> [delete-token] Although only local, don't you want to sign this somehow?

`delete-token` was removed from the standard in the last commit

> Daemon in one but not in other

`delete-token` was removed, but the reason for get-daemon-tokens over get-tokens is that get-tokens is ambiguous with regards to other functionality in the API.

> A HTTP 400 code for invalid address is not what I would expect from a REST perspective. 

This means the factoid address submitted with `fa-address` was malformed, or 400 - bad request(malformed request). Should this be clarified in the spec better? 404 doesn't quite make sense for these purposes. A "not found"(unused) valid Factoid address in FAT has a balance of 0, just like in Factom.

>  I would suggest HTTP 409 for Token Syncing

Totally right, 409 is a better fit for this case for a myriad of reasons.  Thanks!

408 implies some "turning of the gears" and giving up after sometime, however a response in the case of syncing would return almost immediately.

PaulBernier commented on fatips/300.md:
```diff
@@ -0,0 +1,438 @@
+
+## Daemon Methods
+
+### `get-daemon-tokens`:
```
`get-tracked-tokens` would be a more explicit name.

drkatz replied:
I like that :+1:

PaulBernier replied:
What about the method to start tracking a new token?

drkatz replied:
Currently calling a token method of an untracked token will begin tracking it. The API will respond with RPC Error Code `-32805 - Token Syncing` until it's synced and ready to serve requests. This suffices as an equivalent to the suggested call in my eyes, as you can just call a token method fire&forget for the same purpose. Thoughts?

PaulBernier replied:
Ok I think it's good

PaulBernier commented on fatips/300.md:
```diff
@@ -0,0 +1,438 @@
+
+| Name          | Type   | Description                             | Validation                                 | Required |
+| ------------- | ------ | --------------------------------------- | ------------------------------------------ | -------- |
+| `nf-token-id` | string | The unique ID of the non fungible token | Must be a valid issued non fungible token. | Y        |
```
there is no need for the chain/issuer here? The tokens ID are not going to be globally unique, are they?

drkatz replied:
The issuer root chain ID + tokenID or chain ID are required for all token methods (see the top section of token methods header)

PaulBernier replied:
Hm ok that's a bit disconcerting. I guess if we consider it is a spec and not a doc it may be fine. But as a developer I'd certainly jump right to the API I want to learn about without checking all the preamble.

drkatz replied:
Thanks for the input, we can always add it back later. It's a bit more maintainable to have it it one place for now until we finalize things :+1:

nklomp commented:
> Because it already has to be known to make the call in the first place 😉
Yes, but if you are doing a lot of async processing having the correlation in the response itself is handy ;)

>why can't I control pagination myself?
>Could you please explain how you are limited in your control of pagination with the current API?
I would like to specify the amount of pagination but was probably looking at an older version ;)

AdamSLevy commented:
Regarding error code -32801 https://github.com/Factom-Asset-Tokens/FAT/blob/FATIP-300-FAT-RPC-API-Standard/fatips/300.md#-32801---invalid-token

How would this "invalid issuance" situation occur? At the moment the spec for fat-0 allows for multiple invalid issuance to occur prior to the single accepted issuance entry. My implementation simply ignored invalid issuance entries. So is this a token chain that exists and is tracked but is not yet issued?

AdamSLevy commented:
Regarding error code -32801 https://github.com/Factom-Asset-Tokens/FAT/blob/FATIP-300-FAT-RPC-API-Standard/fatips/300.md#-32801---invalid-token

How would this "invalid issuance" situation occur? At the moment the spec for fat-0 allows for multiple invalid issuance to occur prior to the single accepted issuance entry. My implementation simply ignored invalid issuance entries. So is this a token chain that exists and is tracked but is not yet issued?

AdamSLevy commented:
Regarding error code -32801 https://github.com/Factom-Asset-Tokens/FAT/blob/FATIP-300-FAT-RPC-API-Standard/fatips/300.md#-32801---invalid-token

How would this "invalid issuance" situation occur? At the moment the spec for fat-0 allows for multiple invalid issuance to occur prior to the single accepted issuance entry. My implementation simply ignored invalid issuance entries. So is this a token chain that exists and is tracked but is not yet issued?

PaulBernier commented on fatips/300.md:
```diff
@@ -0,0 +1,438 @@
+| Name     | Type   | Description                                                  | Validation                                                   | Required |
+| -------- | ------ | ------------------------------------------------------------ | ------------------------------------------------------------ | -------- |
+| `tx`     | object | The fully formed & signed FAT transaction JSON               | Must conform to all transaction validation criteria of the token's spec | Y        |
+| `extids` | array  | The array of hex encoded ed25519 signatures & RCDs (extids) for each transaction input | Must be valid Signatures & RCDS as per the transaction implementation of the token spec. | Y        |
```
This parameter shouldn't be named extids. This is a common pitfall to name things based on how they are going to be used instead of what they are doing (I wish I could insert a relevant article here but I cannot find one rn). You didn't name the first parameter `content`. At the API level you should be aware of how this data is going to be used. So it should be name something like "signaturesRcds" or along those lines

drkatz replied:
Thanks @PaulBernier, good point

drkatz commented:
> Regarding error code -32801 https://github.com/Factom-Asset-Tokens/FAT/blob/FATIP-300-FAT-RPC-API-Standard/fatips/300.md#-32801---invalid-token
> 
> How would this "invalid issuance" situation occur? At the moment the spec for fat-0 allows for multiple invalid issuance to occur prior to the single accepted issuance entry. My implementation simply ignored invalid issuance entries. So is this a token chain that exists and is tracked but is not yet issued?

 I think the invalid issuance would occurr if the token chain had been established, but did not contain a valid issuance(s)

PaulBernier commented:
> There is an issue with send-transaction that I see. When you submit a signature I believe it is best practice to submit the raw signed data as well. So rather than submit the valid tx JSON, the user should instead provide the raw hex data that represents the exact JSON structure signed by the signatures.
> 
> If someone wants to encode a tx with extra whitespace that is fine but we need the exact data to verify the signature.
> 
> On the backend this data will get converted to the raw bytes the hex represents, and then unmarshalled into the concrete type and checked for validity, all prior to submitting the entry with the exact content data the user supplied.

Have we already debated the option of fatd integrating with walletd and handling the signing? (the same way walletd do the signing when using methods such as `compose-entry`). Walletd can hold EC keys, FCT keys and even identity keys, all of which we need when dealing with FAT tokens, so having fatd work with walletd doesn't sound too far fetched.

Otherwise in the current situation the client of the daemon that want to call `send-transaction` is almost building all the entry itself and then `send-transaction` just forward it to factomd, with some extra validation on top.

drkatz commented:
**NOTICE**
This standard has been deprecated in favor of [fatd](https://github.com/Factom-Asset-Tokens/fatd) specifying it's own [RPC documentation](https://github.com/Factom-Asset-Tokens/fatd/blob/develop/RPC.md). The RPC API will not be standardized, and left up to each implementation.
